# Ассоциативные правила

In [1]:
import pandas as pd
import numpy as np
from apyori import apriori

In [2]:
df = pd.read_csv('datasets/Market_Basket_Optimisation.csv')
df

shrimp            almonds      avocado    vegetables mix  \
0            burgers          meatballs         eggs               NaN   
1            chutney                NaN          NaN               NaN   
2             turkey            avocado          NaN               NaN   
3      mineral water               milk   energy bar  whole wheat rice   
4     low fat yogurt                NaN          NaN               NaN   
...              ...                ...          ...               ...   
7495          butter         light mayo  fresh bread               NaN   
7496         burgers  frozen vegetables         eggs      french fries   
7497         chicken                NaN          NaN               NaN   
7498        escalope          green tea          NaN               NaN   
7499            eggs    frozen smoothie  yogurt cake    low fat yogurt   

     green grapes whole weat flour yams cottage cheese energy drink  \
0             NaN              NaN  NaN            NaN          NaN   
1             NaN              NaN  NaN            NaN          NaN   
2             NaN              NaN  NaN            NaN          NaN   
3       green tea              NaN  NaN            NaN          NaN   
4             NaN              NaN  NaN            NaN          NaN   
...           ...              ...  ...            ...          ...   
7495          NaN              NaN  NaN            NaN          NaN   
7496    magazines        green tea  NaN            NaN          NaN   
7497          NaN              NaN  NaN            NaN          NaN   
7498          NaN              NaN  NaN            NaN          NaN   
7499          NaN              NaN  NaN            NaN          NaN   

     tomato juice low fat yogurt green tea honey salad mineral water salmon  \
0             NaN            NaN       NaN   NaN   NaN           NaN    NaN   
1             NaN            NaN       NaN   NaN   NaN           NaN    NaN   
2             NaN            NaN       NaN   NaN   NaN           NaN    NaN   
3             NaN            NaN       NaN   NaN   NaN           NaN    NaN   
4             NaN            NaN       NaN   NaN   NaN           NaN    NaN   
...           ...            ...       ...   ...   ...           ...    ...   
7495          NaN            NaN       NaN   NaN   NaN           NaN    NaN   
7496          NaN            NaN       NaN   NaN   NaN           NaN    NaN   
7497          NaN            NaN       NaN   NaN   NaN           NaN    NaN   
7498          NaN            NaN       NaN   NaN   NaN           NaN    NaN   
7499          NaN            NaN       NaN   NaN   NaN           NaN    NaN   

     antioxydant juice frozen smoothie spinach  olive oil  
0                  NaN             NaN     NaN        NaN  
1                  NaN             NaN     NaN        NaN  
2                  NaN             NaN     NaN        NaN  
3                  NaN             NaN     NaN        NaN  
4                  NaN             NaN     NaN        NaN  
...                ...             ...     ...        ...  
7495               NaN             NaN     NaN        NaN  
7496               NaN             NaN     NaN        NaN  
7497               NaN             NaN     NaN        NaN  
7498               NaN             NaN     NaN        NaN  
7499               NaN             NaN     NaN        NaN  

[7500 rows x 20 columns]

Заполняем пропущенные значения:

In [3]:
df = df.ffill(axis=1)
df

shrimp            almonds         avocado    vegetables mix  \
0            burgers          meatballs            eggs              eggs   
1            chutney            chutney         chutney           chutney   
2             turkey            avocado         avocado           avocado   
3      mineral water               milk      energy bar  whole wheat rice   
4     low fat yogurt     low fat yogurt  low fat yogurt    low fat yogurt   
...              ...                ...             ...               ...   
7495          butter         light mayo     fresh bread       fresh bread   
7496         burgers  frozen vegetables            eggs      french fries   
7497         chicken            chicken         chicken           chicken   
7498        escalope          green tea       green tea         green tea   
7499            eggs    frozen smoothie     yogurt cake    low fat yogurt   

        green grapes whole weat flour            yams  cottage cheese  \
0               eggs             eggs            eggs            eggs   
1            chutney          chutney         chutney         chutney   
2            avocado          avocado         avocado         avocado   
3          green tea        green tea       green tea       green tea   
4     low fat yogurt   low fat yogurt  low fat yogurt  low fat yogurt   
...              ...              ...             ...             ...   
7495     fresh bread      fresh bread     fresh bread     fresh bread   
7496       magazines        green tea       green tea       green tea   
7497         chicken          chicken         chicken         chicken   
7498       green tea        green tea       green tea       green tea   
7499  low fat yogurt   low fat yogurt  low fat yogurt  low fat yogurt   

        energy drink    tomato juice  low fat yogurt       green tea  \
0               eggs            eggs            eggs            eggs   
1            chutney         chutney         chutney         chutney   
2            avocado         avocado         avocado         avocado   
3          green tea       green tea       green tea       green tea   
4     low fat yogurt  low fat yogurt  low fat yogurt  low fat yogurt   
...              ...             ...             ...             ...   
7495     fresh bread     fresh bread     fresh bread     fresh bread   
7496       green tea       green tea       green tea       green tea   
7497         chicken         chicken         chicken         chicken   
7498       green tea       green tea       green tea       green tea   
7499  low fat yogurt  low fat yogurt  low fat yogurt  low fat yogurt   

               honey           salad   mineral water          salmon  \
0               eggs            eggs            eggs            eggs   
1            chutney         chutney         chutney         chutney   
2            avocado         avocado         avocado         avocado   
3          green tea       green tea       green tea       green tea   
4     low fat yogurt  low fat yogurt  low fat yogurt  low fat yogurt   
...              ...             ...             ...             ...   
7495     fresh bread     fresh bread     fresh bread     fresh bread   
7496       green tea       green tea       green tea       green tea   
7497         chicken         chicken         chicken         chicken   
7498       green tea       green tea       green tea       green tea   
7499  low fat yogurt  low fat yogurt  low fat yogurt  low fat yogurt   

     antioxydant juice frozen smoothie         spinach       olive oil  
0                 eggs            eggs            eggs            eggs  
1              chutney         chutney         chutney         chutney  
2              avocado         avocado         avocado         avocado  
3            green tea       green tea       green tea       green tea  
4       low fat yogurt  low fat yogurt  low fat yogurt  low fat yogurt  
...                ...             ...      

Создадим список транзакций, перебирая строки и столбцы и преобразуя значения в string. 

In [4]:
transactions = []
for i in range(0, 7500): 
    transactions.append([str(df.values[i,j]) for j in range(0, 20)])

Выполняем ассоциативный анализ правил на транзакциях, установив параметры: 

In [5]:
result = list(apriori(transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 4, min_length = 2))

In [6]:
association_results = list(result) #преобразовали вывод алгоритма в список с именем

Выводим длину списка:

In [7]:
print(len(result))

23


23 ассоциативных правила на основе анализа данных.

Выводим первый элемент из списка result:

In [8]:
print(result[0])

RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004533333333333334, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.2905982905982906, lift=4.843304843304844)])


items(предметы): {'light cream', 'chicken'}

Support(поддержка): 0.00453 (это доля транзакций, в которых встречаются 'light cream' и 'chicken')

Ordered_statistics(статистика_порядка): содержит информацию о правиле 'light cream' -> 'chicken'

Confidence(уверенность): 0.2906 (это вероятность, что 'chicken' будет куплено, если куплен 'light cream')

Lift(поддержка): 4.8433 (это показатель, который говорит о том, насколько более вероятно купить 'chicken', если также куплен 'light cream', по сравнению со случаем, когда они покупаются независимо)

Создаем DataFrame с использованием результатов ассоциативного анализа и сохраняем его в переменной nm:

In [9]:
rules=[]
for item in result:

    # первый индекс внутреннего списка
    # Содержит базовый предмет и добавляет предмет
    pair = item[0] 
    items = [x for x in pair]
    v1=str(items[0]) 
    v2= str(items[1])

    #второй индекс внутреннего списка
    v3= str(item[1])

    #третий индекс списка, расположенный на 0-м месте
    #третьего индекса внутреннего списка

    v4= str(item[2][0][2])
    v5=str(item[2][0][3])
    rows=(v1,v2,v3,v4,v5)
    rules.append(rows)
label=['title1','title2','suppot','confidence','lift']  
nm=pd.DataFrame.from_records(rules,columns=label)
nm

title1             title2                 suppot  \
0         light cream            chicken   0.004533333333333334   
1            escalope              pasta   0.005866666666666667   
2               honey      fromage blanc  0.0033333333333333335   
3           olive oil  whole wheat pasta                  0.008   
4              shrimp              pasta   0.005066666666666666   
5            tomatoes  frozen vegetables  0.0030666666666666668   
6             cereals        ground beef  0.0030666666666666668   
7       herb & pepper        ground beef                  0.004   
8                eggs      herb & pepper  0.0041333333333333335   
9        french fries      herb & pepper                 0.0032   
10      herb & pepper        ground beef                 0.0064   
11       tomato sauce        ground beef  0.0030666666666666668   
12               soup          olive oil                 0.0036   
13               soup           tomatoes  0.0030666666666666668   
14          olive oil  whole wheat pasta  0.0038666666666666667   
15  frozen vegetables             shrimp                 0.0032   
16      mineral water    frozen smoothie                 0.0032   
17      mineral water  frozen vegetables  0.0037333333333333333   
18      mineral water  frozen vegetables                 0.0044   
19          olive oil  frozen vegetables  0.0033333333333333335   
20               soup  frozen vegetables  0.0030666666666666668   
21  frozen vegetables             shrimp  0.0033333333333333335   
22      mineral water           tomatoes  0.0030666666666666668   

             confidence               lift  
0    0.2905982905982906  4.843304843304844  
1   0.37288135593220345  4.700185158809287  
2    0.2450980392156863  5.178127589063795  
3    0.2714932126696833  4.130221288078346  
4    0.3220338983050848  4.514493901473151  
5    0.2987012987012987  4.366978051188577  
6   0.45999999999999996  4.681139755766621  
7    0.4411764705882353  4.489584164737808  
8   0.20666666666666667  4.177897574123989  
9   0.23076923076923078  4.665146174580137  
10   0.3934426229508197  4.003825878061259  
11   0.2169811320754717  5.535232961108973  
12  0.21093749999999997  4.174224934036939  
13  0.21904761904761905  4.334715416509612  
14   0.4027777777777778  6.127450980392156  
15  0.32876712328767127  4.608884905901934  
16  0.20512820512820515  4.273504273504273  
17  0.22047244094488186  4.593175853018372  
18  0.25984251968503935  4.350042182227221  
19  0.29411764705882354  6.127450980392157  
20  0.38333333333333336  7.986111111111112  
21             0.390625  4.097465034965035  
22  0.26136363636363635  4.375507305194805

Сортируем DataFrame nm по столбцу 'lift' в порядке убывания и выводим первые 20 строк. Это позволит увидеть топ-20 ассоциативных правил по показателю поддержки.


In [10]:
nm.sort_values(by=['lift'], ascending=False).head(20)

title1             title2                 suppot  \
20               soup  frozen vegetables  0.0030666666666666668   
19          olive oil  frozen vegetables  0.0033333333333333335   
14          olive oil  whole wheat pasta  0.0038666666666666667   
11       tomato sauce        ground beef  0.0030666666666666668   
2               honey      fromage blanc  0.0033333333333333335   
0         light cream            chicken   0.004533333333333334   
1            escalope              pasta   0.005866666666666667   
6             cereals        ground beef  0.0030666666666666668   
9        french fries      herb & pepper                 0.0032   
15  frozen vegetables             shrimp                 0.0032   
17      mineral water  frozen vegetables  0.0037333333333333333   
4              shrimp              pasta   0.005066666666666666   
7       herb & pepper        ground beef                  0.004   
22      mineral water           tomatoes  0.0030666666666666668   
5            tomatoes  frozen vegetables  0.0030666666666666668   
18      mineral water  frozen vegetables                 0.0044   
13               soup           tomatoes  0.0030666666666666668   
16      mineral water    frozen smoothie                 0.0032   
8                eggs      herb & pepper  0.0041333333333333335   
12               soup          olive oil                 0.0036   

             confidence               lift  
20  0.38333333333333336  7.986111111111112  
19  0.29411764705882354  6.127450980392157  
14   0.4027777777777778  6.127450980392156  
11   0.2169811320754717  5.535232961108973  
2    0.2450980392156863  5.178127589063795  
0    0.2905982905982906  4.843304843304844  
1   0.37288135593220345  4.700185158809287  
6   0.45999999999999996  4.681139755766621  
9   0.23076923076923078  4.665146174580137  
15  0.32876712328767127  4.608884905901934  
17  0.22047244094488186  4.593175853018372  
4    0.3220338983050848  4.514493901473151  
7    0.4411764705882353  4.489584164737808  
22  0.26136363636363635  4.375507305194805  
5    0.2987012987012987  4.366978051188577  
18  0.25984251968503935  4.350042182227221  
13  0.21904761904761905  4.334715416509612  
16  0.20512820512820515  4.273504273504273  
8   0.20666666666666667  4.177897574123989  
12  0.21093749999999997  4.174224934036939

Находим правила, представляющие интерес. Усатановим более жесткие пороги для параметров ассоциативного анализа, такие как confidence(уверенность) и support(поддержка). Увеличим значения min_support и min_confidence.

In [11]:
result = list(apriori(transactions, min_support=0.005,min_confidence=0.3, min_lift=4, min_length=2))
result

[RelationRecord(items=frozenset({'escalope', 'pasta'}), support=0.005866666666666667, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.37288135593220345, lift=4.700185158809287)]),
 RelationRecord(items=frozenset({'shrimp', 'pasta'}), support=0.005066666666666666, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'shrimp'}), confidence=0.3220338983050848, lift=4.514493901473151)]),
 RelationRecord(items=frozenset({'herb & pepper', 'ground beef', 'spaghetti'}), support=0.0064, ordered_statistics=[OrderedStatistic(items_base=frozenset({'herb & pepper', 'spaghetti'}), items_add=frozenset({'ground beef'}), confidence=0.3934426229508197, lift=4.003825878061259)])]

Из этих правил с товарами 'spaghetti', 'herb & pepper' и 'ground beef' имеет наивысший лифт (4.0038), что может считаться наиболее интересным с точки зрения ассоциации между товарами

Нахождение правила с максимальным лифтом:

In [12]:
max_lift_rule = nm.sort_values(by='lift', ascending=False).iloc[0]
max_lift_rule

title1                         soup
title2            frozen vegetables
suppot        0.0030666666666666668
confidence      0.38333333333333336
lift              7.986111111111112
Name: 20, dtype: object

Найти товары, которые покупают вместе с пастой:

In [13]:
greeting_card_rules = nm[nm['title1'] == 'pasta']
related_items = greeting_card_rules[['title2', 'confidence', 'lift']].sort_values(by='lift', ascending=False)
related_items

Empty DataFrame
Columns: [title2, confidence, lift]
Index: []